### PR TITLE
(FACT-2738) Write acceptance test that compares Facter 3 and Facter 4 output

### PR DESCRIPTION
**Pull request by IrimieBogdan**
Compare Facter 3 with Facter 4 output. We are able to exclude facts via the `EXCLUDE_LIST `. The excluded fact name can also be a regex because there are fact names that are built dynamically.

### CODE REVIEW

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

File: acceptance/lib/puppet/acceptance/fact_dif.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      @diff[fact_path.join('.')] = { new_value: new_value.inspect, old_value: old_value.inspect }
+    end
+  end
+
+  def excluded?(fact_name)
+    EXCLUDE_LIST.each do |excluded_fact|
+      return true if fact_name =~ /#{excluded_fact}/
+    end
+
+    false
```
Comment (gimmyxd) [inline acceptance/lib/puppet/acceptance/fact_dif.rb:64]: you can replace this with
```ruby
EXCLUDE_LIST.any? {|el| fact_name =~ /#{el}/} 
```

File: acceptance/tests/ensure_facter_3_and_facter_4_output_matches.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    end
+
+    step 'compare Facter 3 to Facter 4 outputs' do
+      fact_dif = FactDif.new(@facter3_output, @facter4_output)
+      unless fact_dif.difs.empty?
+        fail_test("Facter 3 and Facter 4 outputs have the fallowing differences:  #{fact_dif.difs}")
+      end
+    end
+
+    teardown do
```
Comment (gimmyxd) [inline acceptance/tests/ensure_facter_3_and_facter_4_output_matches.rb:29]: teardown should be registered as early as possible

File: acceptance/lib/puppet/acceptance/fact_dif.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      memoryfree_mb swapsize_mb swapfree_mb partitions./dev/mapper/vglocalhost-swap_1.size_bytes
+      partitions./dev/mapper/vglocalhost-root.size_bytes partitions./dev/sda1.size_bytes clientnoop
+      memory.system.available memoryfree system_uptime.uptime uptime facterversion lsbmajdistrelease blockdevices
+      filesystems hypervisors.vmware.version memory.swap.capacity sshfp_ecdsa sshfp_rsa sshfp_ed25519 sshfp_dsa
+      blockdevice_.*_vendor blockdevice_.*_model blockdevice_.*_size mountpoints.* partitions.* operatingsystemrelease
+      os.release.full mtu_.* networking.interfaces.* scope6 disks.* hypervisors.lpar.partition_number.* interfaces.*
+      memory.* processors.speed serialnumber sshdsakey sshrsakey swapfree hypervisors.xen.privileged os.release.minor
+      boardassettag dmi.board.asset_tag boardassettag dmi.board.asset_tag lsbdistrelease lsbminordistrelease
+      processor.* processors.models.* processors.speed bios_release_date bios_vendor bios_version chassisassettag
+      chassistype dmi.* hypervisors.zone.* manufacturer productname virtual zones os.distro.description kernelmajversion
```
Comment (gimmyxd) [inline acceptance/lib/puppet/acceptance/fact_dif.rb:19]: Would be safer to have the facts that are composed to explicitly contain the `.`?
For example 
- `/dmi.*/` will also exclude `dmix`
- `/dmi\..*/` will only exclude `dmi.` constructs

Not sure if there are any overlapping facts that can get accidentally excluded.

Comment (IrimieBogdan) [reply acceptance/lib/puppet/acceptance/fact_dif.rb]: Changed this for all facts, except the ones for which we have `.` and  legacy fact notation that start with the same pattern.

File: acceptance/tests/ensure_facter_3_and_facter_4_output_matches.rb (hunk 2 above)
Comment (IrimieBogdan) [reply acceptance/tests/ensure_facter_3_and_facter_4_output_matches.rb]: moved it to the top of the block

**Comment by IrimieBogdan** [comment]
Moved the `EXCLUDE_LIST ` in the acceptance test to increase visibility of the excluded facts.

**Comment by IrimieBogdan** [comment]
Jenkins please test this on all

**Pull request merged**